### ISSUE TITLE
coff write causes multiply defined __real@3ff0000000000000

### ISSUE THREAD

**Issue by jayvdb** [gimli-rs/object#159]
In the same setup as https://github.com/gimli-rs/object/issues/149 , Cythonised module [`gevent._queue`](https://github.com/gevent/gevent/blob/master/src/gevent/_queue.pxd) on Windows parses and writes without error, but then I encounter linker warning when linking it in with libpythonXY.a

```console
"C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\lib.exe" "-out:C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\pyoxidizer\\libpythonXY.a" "-nologo" "@C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\pyoxidizer\\libpythonXY.a.args"
...
gevent._queue.0.o : warning LNK4006: __real@3ff0000000000000 already defined in gevent.libuv._corecffi.29.o; second definition ignored
```
And then an error when generating the final binary
```console
C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.2viwg4n27f4k3ugl.rcgu.o" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.39xgc2919xpkdbm2.rcgu.o" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.47qx6sds81fagm4s.rcgu.o" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.4tpwh6odkuci599r.rcgu.o" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.4w5ugkvi6xt6vtkj.rcgu.o" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.9zz25lg0q0qieuh.rcgu.o" "/OUT:C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.exe" "C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\pyapp.4tawtyk1l821zf45.rcgu.o" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps" "/LIBPATH:C:/Users/VssAdministrator/pyapp\\build\\target\\debug\\deps" "/LIBPATH:C:/Users/VssAdministrator/pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\pyoxidizer" "/LIBPATH:C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libpyembed-4b48f565d7a5da92.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libuuid-daedd81f20ecf0e1.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand-d1e04b03716ec118.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_xorshift-ea2a037dc87488be.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_pcg-b02ce1ff882f9fe6.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_hc-acc7902f7cbc2f51.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_chacha-2281fa3a7851e606.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_isaac-ecc623c48d28a16d.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_core-c14dd5f160e54f1c.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_os-edfe7944a02c998d.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_jitter-1664950d1b77a195.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libwinapi-54541c97a078f74e.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\librand_core-c4447b557c2a0676.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\liblazy_static-9ea7c902478200db.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libcpython-e576a5f71afb40bd.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libnum_traits-c0b182d52757fad1.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libbyteorder-0cbdb9b2f37557d2.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libpython3_sys-91dda59b980cbd2f.rlib" "C:\\Users\\VssAdministrator\\pyapp\\build\\target\\x86_64-pc-windows-msvc\\debug\\deps\\liblibc-516b44f1ea4d5be3.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libstd-d2a75bc74b11e2cc.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libpanic_unwind-915bfcc230c4eda6.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libhashbrown-2d959e942a9e3e5c.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-d889c9534777e0f9.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libbacktrace-c0632338fc51212a.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_demangle-c23d6f351d09fb94.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libunwind-2098d630f29dcc07.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcfg_if-68cf5df90ef019f9.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liblibc-744c9ad3b176c572.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\liballoc-f60261e32bf289cf.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\librustc_std_workspace_core-2ef81bd49a7650fd.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcore-5779894fffb6f902.rlib" "C:\\Rust\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-7b6566d5f1691e05.rlib" "Ole32.lib" "OleAut32.lib" "User32.lib" "cabinet.lib" "msi.lib" "msvcrt.lib" "rpcrt4.lib" "shlwapi.lib" "version.lib" "winmm.lib" "advapi32.lib" "crypt32.lib" "gdi32.lib" "iphlpapi.lib" "libcrypto.lib" "libssl.lib" "psapi.lib" "shell32.lib" "user32.lib" "userenv.lib" "ws2_32.lib" "advapi32.lib" "credui.lib" "kernel32.lib" "secur32.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "pythonXY.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "msvcrt.lib"
...
          libpyembed-4b48f565d7a5da92.rlib(gevent._queue.0.o) : error LNK2005: __real@3ff0000000000000 already defined in libpyembed-4b48f565d7a5da92.rlib(longobject.obj)
```
where longobject.obj is from [Cython](https://github.com/python/cpython/blob/530f506ac91338b55cf2be71b1cdf50cb077512f/Objects/boolobject.c)

Inside `object` the symbol prints as

> Symbol { name: Some("__real@3ff0000000000000"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(8)), undefined: false, weak: false, scope: Linkage }

sections according to objdump

```console
build\temp.win-amd64-3.7\Release\src/gevent/queue.obj:     file format pe-x86-64

    Sections:
    Idx Name          Size      VMA               LMA               File off  Algn
      0 .drectve      00000060  0000000000000000  0000000000000000  000001a4  2**0
                      CONTENTS, READONLY, DEBUGGING, EXCLUDE, NOREAD
      1 .debug$S      000000c8  0000000000000000  0000000000000000  00000204  2**0
                      CONTENTS, READONLY, DEBUGGING
      2 .text$mn      00025783  0000000000000000  0000000000000000  000002cc  2**4
                      CONTENTS, ALLOC, LOAD, RELOC, READONLY, CODE
      3 .data         00007c71  0000000000000000  0000000000000000  00038ed3  2**4
                      CONTENTS, ALLOC, LOAD, RELOC, DATA
      4 .bss          00001110  0000000000000000  0000000000000000  00000000  2**4
                      ALLOC
      5 .rdata        00000dea  0000000000000000  0000000000000000  000427dc  2**4
                      CONTENTS, ALLOC, LOAD, READONLY, DATA
      6 .xdata        00001710  0000000000000000  0000000000000000  000435c6  2**2
                      CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
      7 .pdata        00001098  0000000000000000  0000000000000000  00045b5e  2**2
                      CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
      8 .rdata        00000008  0000000000000000  0000000000000000  00049572  2**3
                      CONTENTS, ALLOC, LOAD, READONLY, DATA, LINK_ONCE_DISCARD (COMDAT __real@3ff0000000000000 2385)
      9 .chks64       00000050  0000000000000000  0000000000000000  0004957a  2**2
                      CONTENTS, READONLY, DEBUGGING, EXCLUDE, NOREAD
```

From what I read, these `__real@..` symbols are supposed to be weak, and that `LINK_ONCE_DISCARD` in the section header suggests the same thing.

objdump -t ..
```console
...
    [2382](sec  0)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 __ImageBase
    [2383](sec  9)(fl 0x00)(ty   0)(scl   3) (nx 1) 0x0000000000000000 .rdata
    AUX scnlen 0x8 nreloc 0 nlnno 0 checksum 0xa2dacc80 assoc 0 comdat 2
    [2385](sec  9)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 __real@3ff0000000000000
    [2386](sec  0)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 __security_cookie
    [2387](sec  0)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x0000000000000000 _fltused
    [2388](sec 10)(fl 0x00)(ty   0)(scl   3) (nx 1) 0x0000000000000000 .chks64
    AUX scnlen 0x50 nreloc 0 nlnno 0
```

fwiw, these three problems with "gevent" are likely to be the main batch of problems related to PyOxidizer for quite a while.  It only needs to rewrite symbols where two Python packages have used the same name for their module and my analysis of ~2000 packages indicates this is fairly rare, especially when considering which packages PyOxidizer is likely to support in the near future, and the problem only arises if a PyOxidizer user wants to use both in the same project.  I've tested `object` rewriting many of those conflicts without problem, except for Cython-generated modules like gevent.  gevent is especially problematic because it uses the same name as a Python core standard library, and it is a dependency of so many other packages.  For many cases, the Python package authors would be happy to rename their DSO to avoid the conflict, as its name is quite irrelevant to their Python userbase.  This is why I believe pressing ahead with PyOxidizer using `object` is worthwhile - after this initial batch, there will be a managable trickle of 'bugs' in `object`, and in most cases the problem can be avoided by a Python package owner renaming their DSO.
(And PyOxidizer users also have the option of manually renaming the symbols, and I can add support for using `objcopy`, but those should be fallbacks.)

**Comment by jayvdb** [comment]
If I instead rename the symbol in stdlib _queuemodule.obj to avoid the clash, leaving the gevent object unmodified, I get a different set of multiply defined symbols

```console
          libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o) : error LNK2005: "`string'" (??_C@_0BE@KAKMLBKC@can?8t?5allocate?5lock@) already defined in libpyembed-4b48f565d7a5da92.rlib(myreadline.obj)
          libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o) : error LNK2005: "`string'" (??_C@_0BL@DOGKBCNL@timeout?5value?5is?5too?5large@) already defined in libpyembed-4b48f565d7a5da92.rlib(faulthandler.obj)
          libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o) : error LNK2005: "`string'" (??_C@_07DDHNKDGP@timeout@) already defined in libpyembed-4b48f565d7a5da92.rlib(faulthandler.obj)
          libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o) : error LNK2005: "`string'" (??_C@_03HNAFFKGA@get@) already defined in libpyembed-4b48f565d7a5da92.rlib(dictobject.obj)
          libpyembed-4b48f565d7a5da92.rlib(semaphore.obj) : error LNK2005: "`string'" (??_C@_05PIBOEKAB@block@) already defined in libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o)
          libpyembed-4b48f565d7a5da92.rlib(sqlite3.obj) : error LNK2005: "`string'" (??_C@_05PIBOEKAB@block@) already defined in libpyembed-4b48f565d7a5da92.rlib(_queuemodule.obj_stdlib_prefixed.o)
```

These seem to be a different problem, but my blocker is removed if either of these problems are solved.

**Comment by jayvdb** [comment]
My printing of the symbols in the stdlib _queue when rewriting it.
```console
added object symbol name ??_C@_0BE@KAKMLBKC@can?8t?5allocate?5lock@ kind Symbol { name: Some("??_C@_0BE@KAKMLBKC@can?8t?5allocate?5lock@"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(7)), undefined: false, weak: false, scope: Linkage }
2019-11-28T07:21:41.8034616Z added object symbol name ??_C@_0BL@DOGKBCNL@timeout?5value?5is?5too?5large@ kind Symbol { name: Some("??_C@_0BL@DOGKBCNL@timeout?5value?5is?5too?5large@"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(9)), undefined: false, weak: false, scope: Linkage }
2019-11-28T07:21:41.8041185Z added object symbol name ??_C@_07DDHNKDGP@timeout@ kind Symbol { name: Some("??_C@_07DDHNKDGP@timeout@"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(16)), undefined: false, weak: false, scope: Linkage }
2019-11-28T07:21:41.8050906Z added object symbol name ??_C@_03HNAFFKGA@get@ kind Symbol { name: Some("??_C@_03HNAFFKGA@get@"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(26)), undefined: false, weak: false, scope: Linkage }
2019-11-28T07:21:41.8040477Z added object symbol name ??_C@_05PIBOEKAB@block@ kind Symbol { name: Some("??_C@_05PIBOEKAB@block@"), address: 0, size: 0, kind: Data, section_index: Some(SectionIndex(15)), undefined: false, weak: false, scope: Linkage }
```

**Comment by jayvdb** [comment]
This one can be reproduced easily by grabbing https://github.com/indygreg/python-build-standalone/releases/download/20191025/cpython-3.7.5-windows-amd64-20191025T0540.tar.zst (18.2 MB) and using `build/extensions/_queue/_queuemodule.obj`

There is also x86 objects in https://github.com/indygreg/python-build-standalone/releases/download/20191025/cpython-3.7.5-windows-x86-20191025T0549.tar.zst

**Comment by jayvdb** [comment]
I was able to workaround `__real@...` by adding `/FORCE:MULTIPLE` to the linking in PyOxidizer, but then I ran into a different problem with gevent.

So I instead rewrote the symbol in CPython stdlib `_queuemodule.obj`, and I now have a green Windows build (next step is running tests which no doubt will send it red again).

**Comment by philipc** [comment]
I've started looking at this, and I think it'll be best to do this via format-specific section attributes, so a fix for #160 should cover this issue too.
